nservicebus/nhibernate/custom-mappings.md

```diff
@@ -0,0 +1,153 @@
+---
+title: NHibernate Custom mapping with hbm.xml, attributes or fluent api
+summary: Create a custom mapping to change your NHibernate creates the database schema using different techniques.
+tags:
+ - NHibernate
+ - Persistence
+ - Configuration
+ - Mapping
+related:
+ - nservicebus/nhibernate/accessing-data
+ - nservicebus/nhibernate/index
+---
+
+
+Sometimes the default schema that gets generated for your saga does work. For example, you might need to store strings with a different string length, want so use complex types or want to store DateTime values with a high precision or maybe want to tweak the eager or lazy loading to improve performance. You will need to create a custom mapping to achieve this.
+
+Custom mapping options:
+
+* Create a `*.hbm.xml` file for your saga data class
+* Use Fluent NHibernate (popular seperate fluent api)
+* Use NHibernate.Mapping.Attributes
+* Loquacious Configuration (native fluent api)
+
+There are probably even other options but these are the most frequently used.
+
+See http://docs.particular.net/nservicebus/nhibernate/configuration on how you can pass this configuration object to the NServiceBus NHibernate configuration.
+
+NOTE: When you create a custom mapping then you are responsible for the correct behavior assigning the correct primary key.
+
+NOTE: The `[Unique]` attribute will not used to generate a index. You are responsible to add this to your custom mapping. This is especially importent when using optimistic concurrency control and a transaction isolation level different from serializable.
+
+
+## Custom *.hbm.xml mapping
+
+Using NHibernate mapping files is the original way to customize your mapping. It just needs to be created and needs to be embedded as a resource in your assembly or available on the file system.
+
+Look at these pages to see how you can create these mappings:
+
+* http://nhibernate.info/doc/tutorials/first-nh-app/your-first-nhibernate-based-application.html
+
+
+### Mappings as embedded resources
+
+It is not required to create your own NHibernate configuration object and pass it to the NServiceBus NHibernate configuration.
+
+
+### Read mappings from file system instead of embedded resources
+
+If you do not want to embed the `*.hbm.xml` files then you can also load mappings from the file system. Create a custom NHibernate configuration object and use the following example to add mappings from the file system.
+
+```
```

Comment:
suggest you use a snippet for this https://github.com/Particular/docs.particular.net#defining-snippets  let me know if u need help getting started